autogen/oai/gemini.py

```diff
@@ -0,0 +1,316 @@
+
+        if "vision" not in model_name:
+            # A. create and call the chat model.
+            gemini_messages = oai_messages_to_gemini_messages(messages)
+
+            # we use chat model by default
+            model = genai.GenerativeModel(model_name)
+            genai.configure(api_key=self.api_key)
+            chat = model.start_chat(history=gemini_messages[:-1])
+            try:
+                response = chat.send_message(gemini_messages[-1].parts[0].text, stream=stream)
+            except InternalServerError as e:
+                print(e)
+                warnings.warn(
+                    "InternalServerError `500` occurs when calling Gemini's chat model. Retry in 5 seconds...",
+                    UserWarning,
+                )
+                time.sleep(5)
+                return self.create(params)
+            except Exception as e:
+                raise RuntimeError(f"Google GenAI exception occurred while calling Gemini API: {e}")
+            else:
+                # `ans = response.text` is unstable. Use the following code instead.
+                ans: str = chat.history[-1].parts[0].text
+        elif model_name == "gemini-pro-vision":
+            # B. handle the vision model
+            # Gemini's vision model does not support chat history yet
+            model = genai.GenerativeModel(model_name)
+            genai.configure(api_key=self.api_key)
+            # chat = model.start_chat(history=gemini_messages[:-1])
+            # response = chat.send_message(gemini_messages[-1])
+            user_message = oai_content_to_gemini_content(messages[-1]["content"])
+            if len(messages) > 2:
+                warnings.warn(
+                    "Warning: Gemini's vision model does not support chat history yet.",
+                    "We only use the last message as the prompt.",
+                    UserWarning,
+                )
+
+            response = model.generate_content(user_message, stream=stream)
+            # ans = response.text
+            ans: str = response._result.candidates[0].content.parts[0].text
+
+        # 3. convert output
+        message = ChatCompletionMessage(role="assistant", content=ans, function_call=None, tool_calls=None)
+        choices = [Choice(finish_reason="stop", index=0, message=message)]
+
+        prompt_tokens = count_token(params["messages"], model_name)
+        completion_tokens = count_token(ans, model_name)
+
+        response_oai = ChatCompletion(
+            id=str(random.randint(0, 1000)),
+            model=model_name,
+            created=int(time.time() * 1000),
+            object="chat.completion",
+            choices=choices,
+            usage=CompletionUsage(
+                prompt_tokens=prompt_tokens,
+                completion_tokens=completion_tokens,
+                total_tokens=prompt_tokens + completion_tokens,
+            ),
+            cost=0,  # Gemini's cost is zero
```

Comment:
Re. Gemini cost is zero, is this a #TODO for the future? We might want to create a ticket for Gemini costs at $7 / 1 million input tokens; $21 / 1 million output tokens?
https://ai.google.dev/pricing